File: beet/test/compat/enums.ts
import test from 'tape'
import { fixedScalarEnum } from '../../src/beet'
import {
  checkFixedDeserialize,
  checkFixedSerialization,
  checkFixedSerialize,
} from '../utils'
import fixture from './fixtures/enums.json'

enum Directions {
  Up,
  Right,
  Down,
  Left,
}

enum Milligrams {
  Grams,
  Kilograms,
}

function variantFromString<T>(enumType: T & {}, variant: string): T {
  const idx = Object.keys(enumType).indexOf(variant)
  return Object.values(enumType)[idx] as T
}

test('compat enums: directions', (t) => {
  const beet = fixedScalarEnum(Directions)
  for (const { value, data } of fixture.directions) {
    const direction = variantFromString(Directions, value)
    checkFixedSerialization(t, beet, direction, data)
  }
  t.end()
})

test('compat enums: milligrams', (t) => {
  const beet = fixedScalarEnum(Milligrams)
  for (const { value, data } of fixture.milligrams) {
    const milligram = variantFromString(Milligrams, value)
    checkFixedSerialization(t, beet, milligram, data)
  }
  t.end()
})

test('compat enums: directions using int', (t) => {
  const beet = fixedScalarEnum(Directions)
  const buf = Buffer.alloc(beet.byteSize)
  beet.write(buf, 0, Directions.Down)
  // Clearer test to show expected values
  checkFixedSerialize(
    t,
    beet,
    Directions.Down,
    [0x02],
    'Directions.Down Serialize'
  )
  checkFixedDeserialize(
    t,
    beet,
    Directions.Down,
    [0x02],
    'Directions.Down Deserialize'
  )
  t.end()
})

File: tools/bsamples/src/enums.rs
use anyhow::Result;

use borsh::{BorshDeserialize, BorshSerialize};
use serde::{Deserialize, Serialize};

use crate::samples::{produce_samples, Sample};

#[derive(Serialize, Deserialize, BorshSerialize, BorshDeserialize)]
pub enum Direction {
    Up,
    Right,
    Down,
    Left,
}

#[derive(Serialize, Deserialize, BorshSerialize, BorshDeserialize)]
enum Milligrams {
    Grams = 1000,
    Kilograms = 1000_000,
}

#[derive(Serialize, Deserialize)]
pub struct Enums {
    directions: Vec<Sample<Direction>>,
    milligrams: Vec<Sample<Milligrams>>,
}

pub fn produce_enums() -> Result<Enums> {
    let directions: Vec<Sample<Direction>> = produce_samples(vec![Direction::Up, Direction::Down])?;
    let milligrams: Vec<Sample<Milligrams>> =
        produce_samples(vec![Milligrams::Grams, Milligrams::Kilograms])?;

    Ok(Enums {
        directions,
        milligrams,
    })
}
